### ISSUE TITLE
PG doesn't grok Coq's Add LoadPath

### ISSUE THREAD

**Issue by jonleivent** [ProofGeneral/PG#96]
I like the way PG is able to get info from _CoqProject.  But, in those cases where one is just experimenting with a script that isn't yet itself part of a project, it would be nice if PG didn't go bonkers when the script did an Add LoadPath,  Right now, if the current dir has no _CoqProject, and a script in that dir tries to do Add LoadPath to access an existing project (which works when the script is run in CoqIDE, by the way), not only does it fail, but PG pretty much becomes useless after that point - requiring one to kill Emacs and restart.
 
This is Emacs 24.4.1, Coq 8.5pl2, very recent PG build (462e3ef).

**Comment by jonleivent** [comment]
I should point out that the "Add LoadPath" command itself is not failing in PG - it is the subsequent Require or From ... Require based on that loadpath addition that fails, and fails badly.

**Comment by cpitclaudel** [comment]
You're using compile before require, right?

**Comment by jonleivent** [comment]
Yes. 
Also, the targeted module is already compiled.

**Comment by cpitclaudel** [comment]
Can you try disabling compile before require and seeing if that works?

**Comment by jonleivent** [comment]
Yup - it works OK when Compile before Require is disabled.

**Comment by cpitclaudel** [comment]
Ok, makes sense. I'm not familiar with this code; it was written by @hendriktews IIRC.

**Comment by hendriktews** [comment]
I never looked at load path manipulation commands when I worked on the compilation features, because, I believe, at that time, Add Loadpath could not be undone... 

I only have the suggestion to add a file variable for coq-load-path, but then the script does not work outside of PG.

**Comment by jonleivent** [comment]
I guess the whole cross-project logical-physical concept is still not ready for prime time - see for instance: [https://coq.inria.fr/bugs/show_bug.cgi?id=4990](url)

I managed to get PG to respect COQPATH, even when PG is not launched from a shell (I typically launch it from a toolbar) by installing exec-path-from-shell and putting this into my coq-mode-hook:

`(exec-path-from-shell-copy-env "COQPATH")`

But, as outlined in 4990, relying on COQPATH isn't a great option - it's just the only one that seems to currently be supported.

As for coq-load-path, certainly I could set that up as with a file variable, but - yes- the whole point is to get one scheme that works in all tools and utilities.  Hence, I don't think coq-load-path should get used, as it is specifically for PG.  Maybe deprecate it?

**Comment by Matafou** [comment]
Let us try sum this up. We need to understand in what sense the current tools are not sufficient.

In a Coq project there are:

1) Settings that should be shared by all developpers and users and hence should be in a git repo:
    a) coq options related to the project itself:  e.g. -R . MyProject
       ==> Goes in _CoqProject file

    b) Module name of external projects on which compilation depends: e.g. MathComp
       ==> goes nowhere currently, should maybe be compatible with opam syntax, shoudl definitely be shared with all IDEs

2) Settings that should not go in the git repository because they are not shared by all users
    a) ex: the place where the user put OTHER projects in her filesystems
        e.g. /users/home/me/coq/MathComp/x.y
       ==> This can go in a (unversioned) .dir-locals.el file for PG users but then coq_makefile and other IDEs will not work.

Note that coq options related to external projects do not fit in this schema. They can be elaborated from 1b) and 2) by some tools and be added to _CoqProject in some way.

My impression is that we need some kind of _CoqProject.private. It should contain the path information 2a), and _CoqProject would refer to it.

_CoqProject.private:

MathCompPath=/users/me/Coq/MathComp

_CoqProject:

include _CoqProject.private
-R $MathCompPath MathComp
-R . MySubProject
-R ../Foo Foo  


What do you think of this. This i similar to netbeans or eclipse preferences.private stuff.

The only thing we need is the "include" directive to be added and understood by coq_makefile and IDEs. This also looks very like the configure tool but I don't think we want to jump into it do we?

One last note: I think Add LoadPath should not be used at all. Absolute physical path should never appear in coq files IMHO. Is there any reason one would need it instead of -QRI options?

P.

**Comment by Matafou** [comment]
I agree that coq-load-path should not be used too by the way. I think Add Load Path and coq-load-path are to be deprecated, or at least discouraged as soone as 2 coq files are involved.

**Comment by jonleivent** [comment]
I agree that giving _CoqProject the power to include another file is a good way forward here.  But, maybe some limitations on what it can contain should be enforced so that people don't accidentally move options that should be in the project's own _CoqProject into some private include file.

The tool-specific mechanisms like coq-load-path and .dir-locals.el should all be discouraged for such purposes.  $COQPATH should be discouraged because of its inability to handle separate logical vs. physical paths, as well as the difficulty of using env vars with some tools.

The refman should not mention using resource files with Add Load Path as a way to connect projects.  However, there might be other valid uses of Add Load Path - for instance, altering which modules are used for debugging purposes in a single script vs. doing so on a project-wide basis.   Although, maybe if _CoqProject had the ability to have per-file overrides of the -QRI options, that would be a better approach to such things.  I'm thinking here of something like the C/C++ model of project compilation where one often separately compiles debug vs. optimized versions of files and mixes them as needed to debug the project.

**Comment by erikmd** [comment]
Hi,

Le mardi 16 août 2016 à 08:57 -0700, Pierre Courtieu a écrit :
> My impression is that we need some kind of _CoqProject.private. It should contain the path information 2a), and _CoqProject would refer to it.
> 
> _CoqProject.private:
> 
> MathCompPath=/users/me/Coq/MathComp
> 
> _CoqProject:
> 
> include _CoqProject.private
> -R $MathCompPath MathComp
> -R . MySubProject
> -R ../Foo Foo  
> 
> 
> What do you think of this.

Pros: such an "include" syntax would indeed be convenient to setup
used-defined paths for the Coq project's dependencies, without being
tool-specific.

Cons: the main drawback I see is that there is no distinction between
the "internal dependencies" of the project (-R . MySubProject), and
the external dependencies of the project (-R ... MathComp), so for
example if the _CoqProject above would be used with current version of
coq_makefile, "make clean" would remove all MathComp theories as well.

It is true that currently there's no perfect way to setup such
internal/external dependencies for Coq projects.

TTBOMK the best standard way to workaround the issue above consists in
1. installing all external dependencies ("make install") beforehand in
the "user-contrib" subfolder, whose path is hardwired in Coq, then
2. put the project's internal dependencies only in the _CoqProject
file and use coq_makefile

Actually, step 1. above is what is done when the Coq installation is
managed by OPAM, for example when I run:
$ opam install coq-mathcomp-algebra
on my system, all dependencies are installed in a folder such as
~∕.opam/4.02.3+coq8.5/lib/coq/user-contrib

However, maybe the way _CoqProject handle dependencies could be
refined: IIRC doing "coq_makefile -f _CoqProject" amounts to passing
each line of the _CoqProject file to coq_makefile as CLI arguments.
So, coq_makefile could be slightly extended to distinguish between
external dependencies and internal ones: we could add an option "--"
before the list of external dependencies (that should not be erased by
"make clean").

Now I'm realizing that an "include" already exists, in the form of the
flag "-f" !
But I'm also noting that "coq_makefile -R something Foo" raises a
warning if something is not a subfolder of the current directory:
$ coq_makefile -R / Temp >/dev/null
> Warning: in option -R/-Q, / is not a subdirectory of the current directory

As a result, if the option "--" I suggested above was supported in
coq_makefile to allow subsequent -QRI option to deal with arbitrary
folders, we could write the following (without needing to install
MathComp in user-contrib):

> _CoqProject.private:
> 
> MathCompPath=/users/me/Coq/MathComp
> 
> _CoqProject:
> 
> -R . MySubProject
> --
> -f _CoqProject.private
> -R $(MathCompPath) MathComp

What do you think of this?

Erik

**Comment by erikmd** [comment]
Hi,
On second thought, the inclusion of `_CoqProject.private` by `_CoqProject` is not mandatory, as we could write just as well:

> **_CoqProject:**
> -R . MySubProject

> **_CoqProject.private**
> -R /users/me/Coq/MathComp MathComp

and run `$ coq_makefile -f _CoqProject -- -f _CoqProject.private`
(or `$ touch _CoqProject.private && coq_makefile -f _CoqProject -- -f _CoqProject.private`)

which will also work with an empty or non-existing `_CoqProject.private` file &mdash; in the case where the MathComp library has been installed in the `user-contrib` folder...

Best regards,
Erik

**Comment by Matafou** [comment]
IMHO it would be nice not to rely on a fixed  directory. It is not compatible with the fact to support two versions of the same package. I know I know, opam switch blabla, anyway it is less flexible. Let us have a generic way of doing things instead, and make it simple when people use opam.

I like your idea of the double --f. I don't like the separator thing, although I see its interest.
Let us not rely too much on bash. Let us say that -f does nothing if the file does ot exist (just a warning).

**Comment by erikmd** [comment]
Hi Pierre,

> IMHO it would be nice not to rely on a fixed  directory. It is not
> compatible with the fact to support two versions of the same package.

I'm not sure I understood what you mean (please correct me if this is
not the case): you are talking about the path "/users/me/Coq/MathComp"
in the _CoqProject.private example files above, which should be easily
changed to support two versions of the same external dependency.

I agree that such a version change is often necessary to do
(automated) test builds, but at the time PG reads the _CoqProject (or
_CoqProject.private) file the path of the dependency should be "fixed"

… we could try to use env. variables to change dependencies' version,
but IIRC coq_makefile doesn't support bash-like environment $Variables
in _CoqProject files, it partly supports Makefile-like $(Variables)…
so parsing them reliably in PG would be difficult.

So I believe this issue (compiling with various dependencies versions)
should be managed at configure time, or equivalently by a master
Makefile that would call coq_makefile with various options…

> I like your idea of the double --f.

I was proposing to write "-- -f", but why not "--f" indeed :)

> I don't like the separator thing, although I see its interest.

OK. And BTW I've just noticed that with current version
"-f _CoqProject.private" wouldn't work out of the box, because
coq_makefile assumes there is only one "-f _" option.
As a result the Makefile obtained with
> $ coq_makefile -f _CoqProject -f _CoqProject.private -o Makefile
contains
> Makefile: _CoqProject.private
> 	mv -f $@ $@.bak
> 	"$(COQBIN)coq_makefile" -f $< -o $@
> [...]
So indeed, a solution could be to add a "--f" option for a second
(private) project file.

But, if we don't add a "--" option in coq_makefile, maybe some options
"--R", "--Q", "--I" should be added for external dependencies ?

> Let us not rely too much on bash. Let us say that -f does nothing if
> the file does ot exist (just a warning).

Agreed.

**Comment by Matafou** [comment]
I agree to all this.

There are two wishes here:

1) include another (private) file to specify
  + the value of certain physical paths
  + some other private options ?
 and have coq_makefile and pg support this.

2) allow to add/remove options per file and have pg support this syntax.
Meanwhile compile-before-require should maybe try to support Add Load Path?

P



2016-08-16 18:49 GMT+02:00 Jonathan Leivent <notifications@github.com>:

**Comment by erikmd** [comment]
Hi @Matafou and @jonleivent,
I discussed this issue with @silene at ITP'2016, so here is some feedback on this topic:

1. `coq_makefile` does not fully support multiple `-f` options currently; so I've just filed a Coq bugreport at [Bugzilla 5088](https://coq.inria.fr/bugs/show_bug.cgi?id=5088).

2. Regarding external dependencies, the advised method (whether one uses `opam` or not) is to always install the required libraries, by doing `make install` or so beforehand. Otherwise `coq_makefile` does not like very much custom paths passed as -R/-Q arguments.

3. After my suggestion to write something like
   `touch _CoqProject.private && coq_makefile -f _CoqProject -f _CoqProject.private`
   Pierre had suggested to avoid depending on `bash` / `sh`:

   > Let us say that -f does nothing if the file does ot exist (just a warning)

   but doing such a change could disturb the user, and actually depending on `sh` for calling `coq_makefile` is certainly OK, as calling `coq_makefile` assumes the user can use the generated Makefile (so it has at least an `sh`-compatible shell).

Regarding what remains for #96:

I believe the original issue posted by Jonathan does not actually deal with `coq_makefile` nor `_CoqProject`, so if @Matafou / @cpitclaudel  are OK with this I'll rename the current issue to mainly focus on `Add Load Path` and open another issue regarding the parsing of `-f` options from `_CoqProject` files (to sum up, multiple or "nested" `-f` options is not yet supported by `coq_makefile` but fully supported by `coqide`, so that'd be neat to support them in PG as well !)

Best,
Erik

**Comment by jonleivent** [comment]
I am not sure I understand point 2 above:

> Regarding external dependencies, the advised method (whether one uses opam or not) is to always install the required libraries, by doing make install or so beforehand. Otherwise coq_makefile does not like very much custom paths passed as -R/-Q arguments.

When you say "make install or so beforehand", what are you suggesting about their install paths, and how does the client code pick up on those paths?  In my case, both client and library were being developed simultaneously.

**Comment by erikmd** [comment]
Hi Jonathan,

Le vendredi 16 septembre 2016 à 17:21 -0700, Jonathan Leivent a écrit :
> When you say "make install or so beforehand", what are you
> suggesting about their install paths, and how does the client code
> pick up on those paths?
> In my case, both client and library were being developed
> simultaneously.

In my comment I was thinking of relying on the default install path
generated by coq_makefile, namely:
> COQLIBINSTALL="${COQLIB}user-contrib"

The advantage of using this install path is that all "client" projects
will be able to use the libraries installed there without specifying
any -R/-Q arguments for binding *absolute* paths to logical names.

In the use case you mention (concurrent development of two libraries),
I think there are two possible cases:

Case A: both libraries are hosted in the same repository, say in two
subfolders "lib" and "client", and "client" depends on "lib".

In this case we can just use a single _CoqProject file containing e.g.
> -R lib FooLib
> -R client FooClient
> lib/lib.v
> client/client.v

so "coq_makefile -f _CoqProject -o Makefile && make && make install"
will compile everything and put the two libraries in
$COQLIB/user-contrib/FooLib/ and $COQLIB/user-contrib/FooClient/

Case B: the two libraries are hosted in different projects, each of
them containing a _CoqProject (and a Makefile generated similarly)

In this case I would follow the following workflow:
"cd .../lib && make && make install && cd .../client && make"

Does this answer your question?

Kind regards,
Erik

**Comment by jonleivent** [comment]
Up to now, I have not used the Makefile install target at all - instead keeping each part in its own git clone repo.  OK - maybe I have to change...

Interestingly, COQLIB is not documented in the refman as being required to be set by the user - in fact, it is documented quite the opposite way, as only needing to be set to override the position of the standard lib - and even that is only under coqchk in the refman (14.4).  

Also, a coq_makefile generated Makefile does not set COQLIB itself - at least I don't see how it does, based on looking through several Makefiles.  It does consume COQLIB for its install targets, as you point out.  It also consumes, but does not set, DSTROOT, which is not documented at all, and combines it with COQLIB to produce the install path.  Hence, if I have neither DSTROOT nor COQLIB set, the Makefile will behave badly doing installs.

Maybe there is some method to the madness here of using both DSTROOT and COQLIB to create the install path, while having the tools (such as coqchk) use COQLIB to find the standard library.  If so, what is it?

**Comment by erikmd** [comment]
> Interestingly, COQLIB is not documented in the refman as being required to be set by the user - in fact, it is documented quite the opposite way, as only needing to be set to override the position of the standard lib - and even that is only under coqchk in the refman (14.4).

Indeed the refman doesn't say much about COQLIB, but it is not required to be set by the user (see below), and I believe that in most practical cases the user won't need to customize this environment variable.

> Also, a coq_makefile generated Makefile does not set COQLIB itself - at least I don't see how it does, based on looking through several Makefiles. It does consume COQLIB for its install targets, as you point out.

It seems that coq_makefile-generated Makefiles set a default value to COQLIB (and thereby COQLIBINSTALL) thanks to the line `$(call includecmdwithout@,$(COQBIN)coqtop -config)`

To experiment this, you can compare the output of `make printenv` vs.
`make printenv COQLIB=/foo/ DOCDIR=/bar/`

> It also consumes, but does not set, DSTROOT, which is not documented at all, and combines it with COQLIB to produce the install path. Hence, if I have neither DSTROOT nor COQLIB set, the Makefile will behave badly doing installs.

The only documentation I've found on DSTROOT is the comment line in the Makefile
`# DSTROOT to specify a prefix to install path.`
but the installation will behave well even if it is not set, since it will be replaced with the empty string.

> Maybe there is some method to the madness here of using both DSTROOT and COQLIB to create the install path, while having the tools (such as coqchk) use COQLIB to find the standard library. If so, what is it?

I'm not sure if setting COQLIB would have an impact on `coqchk`, AFAICT the COQLIB and DSTROOT environment variables are only useful if one want to override the install path of user-compiled libraries to `$(DSTROOT)/$(COQLIB)/user-contrib/Foo`, by doing something like:
`make install DSTROOT=/prefix/ COQLIB=/foo/`

For example, in my setup the default install path is `~/.opam/4.02.3+coq8.5/lib/coq/user-contrib/Foo`
(which has not to do with Coq' stdlib, that is stored in `~/.opam/4.02.3+coq8.5/lib/coq/theories/`).
